hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNodeRpcServer.java

```diff
@@ -1899,7 +1907,27 @@ private void verifySoftwareVersion(DatanodeRegistration dnReg)
     }
   }
 
-  private static String getClientMachine() {
+  private String getClientMachine() {
+    if (ipProxyUsers != null) {
+      // Get the real user (or effective if it isn't a proxy user)
+      UserGroupInformation user = Server.getRemoteUser().getRealUserOrSelf();
```

Comment:
One more here:
Can this shoot a NPE?
There is a null check below for  ``Server.getRemoteUser()``
```
    String clientMachine = Server.getRemoteAddress();
    if (clientMachine == null) { //not a RPC client
```

Reply:
Nice catch, thank you.

Reply:
Although doing the user operations on the NN without RPC would be unexpected. :)